File: backend/config/firebase.js
/* eslint-disable no-console */
import { initializeApp, cert } from 'firebase-admin/app';
import { getDatabase } from 'firebase-admin/database';
import path from 'path';
import fs from 'fs';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const backendRoot = path.resolve(__dirname, '../');

let serviceAccountPath = path.resolve(backendRoot, 'serviceAccountKey.json');

if (!fs.existsSync(serviceAccountPath)) {
  const files = fs.readdirSync(backendRoot);
  const adminSdkFile = files.find(file => file.includes('firebase-adminsdk') && file.endsWith('.json'));
  if (adminSdkFile) {
    serviceAccountPath = path.resolve(backendRoot, adminSdkFile);
  }
}

let isFirebaseInitialized = false;
let db = null;

if (fs.existsSync(serviceAccountPath)) {
  try {
    const serviceAccount = JSON.parse(fs.readFileSync(serviceAccountPath, 'utf8'));
    
    initializeApp({
      credential: cert(serviceAccount),
      databaseURL: "https://wmhrisdps-default-rtdb.asia-southeast1.firebasedatabase.app"
    });
    
    db = getDatabase();
    isFirebaseInitialized = true;
    console.log('[FIREBASE] Admin SDK initialized successfully.');
  } catch (error) {
    console.error('[FIREBASE] Error initializing Admin SDK:', error.message);
  }
} else {
  console.warn('[FIREBASE] Warning: serviceAccountKey.json not found in backend directory. Real-time signals will not be sent.');
}

export const isInitialized = isFirebaseInitialized;
export const firebaseDb = db;

File: backend/internal/config/firebase.go
package config

import (
	"context"
	"log"
	"os"

	firebase "firebase.google.com/go/v4"
	"firebase.google.com/go/v4/db"
	"google.golang.org/api/option"
)

var (
	FirebaseApp           *firebase.App
	FirebaseDB            *db.Client
	IsFirebaseInitialized bool
)

// InitFirebase initializes the Firebase Admin SDK
func InitFirebase() {
	credentialsPath := os.Getenv("FIREBASE_CREDENTIALS_PATH")
	if credentialsPath == "" {
		log.Println("[FIREBASE] Warning: FIREBASE_CREDENTIALS_PATH not set in .env. Real-time signals will not be sent.")
		return
	}

	dbURL := os.Getenv("FIREBASE_DB_URL")
	if dbURL == "" {
		log.Println("[FIREBASE] Warning: FIREBASE_DB_URL not set in .env. Real-time signals will not be sent.")
		return
	}

	ctx := context.Background()
	opt := option.WithCredentialsFile(credentialsPath)
	conf := &firebase.Config{
		DatabaseURL: dbURL,
	}

	app, err := firebase.NewApp(ctx, conf, opt)
	if err != nil {
		log.Printf("[FIREBASE] Error initializing Admin SDK: %v\n", err)
		return
	}

	database, err := app.Database(ctx)
	if err != nil {
		log.Printf("[FIREBASE] Error initializing RTDB client: %v\n", err)
		return
	}

	FirebaseApp = app
	FirebaseDB = database
	IsFirebaseInitialized = true
	log.Println("[FIREBASE] Admin SDK initialized successfully.")
}
